politeiawww/cmd/politeiavoteraudit/politeiavoteraudit.go

```diff
@@ -0,0 +1,724 @@
+			} else {
+				successCount++
+			}
+		}
+	}
+	if errorCount != 0 {
+		return 0, fmt.Errorf("auditWork: unexpected number of "+
+			"errors: %v", errorCount)
+	}
+
+	if successCount != totalRecords {
+		return 0, fmt.Errorf("auditWork: unexpected totals: "+
+			"got %v want %v", successCount, totalRecords)
+	}
+
+	// if we make it here, all work, despite retries, completed correctly.
+	return totalRecords, nil
+}
+
+func auditFailed(verbose bool, work map[int64][]jsontypes.VoteInterval, success map[int64]jsontypes.BallotResult, failed map[int64]BallotError) (int, error) {
+	// Go through all failed records and ensure thet they are non-fatal
+	// failures.
+	if verbose {
+		fmt.Printf("  Audit: failed\n")
+	}
+
+	totalFailed := 0
+	for _, v := range failed {
+		for _, vv := range v.Ballot.Votes {
+			ok := findSuccess(vv.Ticket, success)
+			if !ok {
+				return 0, fmt.Errorf("permanent error on: %v", vv.Ticket)
+			}
+			totalFailed++
+		}
+	}
+
+	return totalFailed, nil
+}
+
+func auditSuccess(verbose bool, work map[int64][]jsontypes.VoteInterval, success map[int64]jsontypes.BallotResult) (int, error) {
+	if verbose {
+		fmt.Printf("  Audit: success\n")
+	}
+
+	totalSuccess := 0
+	for _, v := range success {
+		ok := findWork(v.Ticket, work)
+		if !ok {
+			return 0, fmt.Errorf("ticket not found in work map: %v", v.Ticket)
+		}
+		totalSuccess++
+	}
+
+	return totalSuccess, nil
+}
+
+func (c *ctx) getVoteResultsReply(token string) (*v1.VoteResultsReply, error) {
+	responseBody, err := c.makeRequest("GET", "/proposals/"+token+"/votes", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var vrr v1.VoteResultsReply
+	err = json.Unmarshal(responseBody, &vrr)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal "+
+			"ProposalVotesReply: %v", err)
+	}
+
+	return &vrr, nil
+}
+
+func auditPi(verbose bool, vrr *v1.VoteResultsReply, work map[int64][]jsontypes.VoteInterval, success map[int64]jsontypes.BallotResult) (int, error) {
+	if verbose {
+		fmt.Printf("  Audit: politeia\n")
+	}
+
+	cvm := make(map[string]*v1.CastVote, len(vrr.CastVotes))
+	for _, cv := range vrr.CastVotes {
+		cvm[cv.Ticket] = &cv
+	}
+
+	votedSuccess := 0
+	for _, s := range success {
+		if _, ok := cvm[s.Ticket]; !ok {
+			continue
+		}
+		votedSuccess++
+	}
+
+	if verbose {
+		fmt.Printf("  Politeia succeful votes: %v\n", votedSuccess)
+	}
+
+	return votedSuccess, nil
+}
+
+func (c *ctx) audit(args []string) error {
+	for k, v := range args {
+		path := filepath.Join(c.cfg.voteDir, v)
+		d, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+
+		work := make(map[int64][]jsontypes.VoteInterval, 1024)  // [timestamp][]VoteInterval
+		success := make(map[int64]jsontypes.BallotResult, 1024) // [timestamp]BallotResult
+		failed := make(map[int64]BallotError, 1024)             // [timestamp]BallotError
+		fmt.Printf("Proposal: %v\n", v)
+
+		vrr, err := c.getVoteResultsReply(v)
+		if err != nil {
+			return err
+		}
+
+		for _, vv := range d {
```

Comment:
This doesn't seem to be able to handle when you have to restart politeiavoter and new journal files get created.  For example, if a vote doesn't get cast in the first set of journal files but does get cast in the second set of journal files, all of the errors from the first set of journal files will still get printed giving the impression that there are uncast votes when in reality all votes have been cast.